ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs

```diff
@@ -471,18 +548,80 @@ completeBlockDownload tracer blockFetchSize inflightlimits
+                   -> m ()
+completeFetchBatch tracer inflightlimits range
+                   FetchClientStateVars {
+                     fetchClientInFlightVar,
+                     fetchClientStatusVar
+                   } = do
+    (inflight, currentStatus) <- atomically $ do
+      inflight <- readTVar fetchClientInFlightVar
+      let !inflight' =
+            assert (if peerFetchReqsInFlight inflight == 1
+                       then peerFetchBytesInFlight inflight == 0
+                         && Set.null (peerFetchBlocksInFlight inflight)
+                       else True)
+            inflight {
+              peerFetchReqsInFlight = peerFetchReqsInFlight inflight - 1
+            }
+      writeTVar fetchClientInFlightVar inflight'
+      currentStatus <- readTVar fetchClientStatusVar
+      return (inflight', currentStatus)
+
+    traceWith tracer $
+      CompletedFetchBatch
+        range
+        inflight inflightlimits
+        currentStatus
+
+
+rejectedFetchBatch :: (MonadSTM m, HasHeader header)
+                   => Tracer m (TraceFetchClientState header)
+                   -> (header -> SizeInBytes)
+                   -> PeerFetchInFlightLimits
+                   -> ChainRange header
+                   -> [header]
                    -> FetchClientStateVars m header
                    -> m ()
-completeFetchBatch tracer FetchClientStateVars {fetchClientInFlightVar} = do
-    atomically $ modifyTVar fetchClientInFlightVar $ \inflight ->
-      assert (if peerFetchReqsInFlight inflight == 1
-                 then peerFetchBytesInFlight inflight == 0
-                   && Set.null (peerFetchBlocksInFlight inflight)
-                 else True)
-      inflight {
-        peerFetchReqsInFlight = peerFetchReqsInFlight inflight - 1
-      }
-    traceWith tracer CompletedFetchBatch
+rejectedFetchBatch tracer blockFetchSize inflightlimits range headers
+                   FetchClientStateVars {
+                     fetchClientInFlightVar,
+                     fetchClientStatusVar
+                   } = do
+    (inflight', currentStatus') <- atomically $ do
+      inflight <- readTVar fetchClientInFlightVar
+      let !inflight' =
+            (deleteHeadersInFlight blockFetchSize headers inflight) {
+              peerFetchReqsInFlight = peerFetchReqsInFlight inflight - 1
+            }
+      writeTVar fetchClientInFlightVar inflight'
+
+      -- Set our status to ready if we're under the low watermark.
+      let currentStatus'
+            | peerFetchBytesInFlight inflight'
+              <= inFlightBytesLowWatermark inflightlimits
+            = PeerFetchStatusReady (peerFetchBlocksInFlight inflight')
+            | otherwise
+            = PeerFetchStatusBusy
+      -- Only update the variable if it changed, to avoid spurious wakeups.
+      currentStatus <- readTVar fetchClientStatusVar
+      when (currentStatus' /= currentStatus) $
+        writeTVar fetchClientStatusVar currentStatus'
```

Comment:
This looks to be copy-pasted from `completeBlockDownload`. Should there be any differences? The two functions share a lot -- factor it out the duplication?

@dcoutts says "copy-paste is OK" on Slack. I'm Approving and leaving the duplication up to him to decide.

Reply:
Ta. I might pull it out into a common STM action.